### PR TITLE
Number of backouts as feature

### LINKED ISSUE #110: Add number of backouts as a feature

### PR DESCRIPTION
**Pull request by ayir**
Fixes #110.

### CODE REVIEW

File: bugbug/repository.py [hunk 1]
```diff
@@ -19,7 +19,7 @@
 COMMITS_DB = 'data/commits.json'
 db.register(COMMITS_DB, 'https://www.dropbox.com/s/mz3afgncx0siijc/commits.json.xz?dl=1')
 
-Commit = namedtuple('Commit', ['node', 'author', 'desc', 'date', 'bug'])
+Commit = namedtuple('Commit', ['node', 'author', 'desc', 'date', 'bug', 'no_of_backouts'])
```
Comment (marco-c) [inline bugbug/repository.py]: I think a given commit can be backed out only once, so this should be a boolean `ever_backedout`

**Comment by marco-c** [comment]
You also need to actually add the feature to bug_features.py, where you will sum the backouts for all commits linked to the bug.

File: bugbug/repository.py (hunk 1 above)
Comment (ayir) [reply bugbug/repository.py]: yes as far as I saw the outputs it was giving only one node.And yes I'll change it.

File: bugbug/bug_features.py [hunk 2]
```diff
@@ -230,6 +230,11 @@ def __call__(self, bug, **kwargs):
         return sum(res) / len(res)
 
 
+class commit_no_of_backouts(object):
+    def __call__(self, bug, **kwargs):
+        return len([commit for commit in bug['commits'] if commit['ever_backedout'] is True])
```
Comment (marco-c) [inline bugbug/bug_features.py]: You can use `sum` to avoid creating a list: `sum(1 for commit in ...)`

File: bugbug/repository.py [hunk 3]
```diff
@@ -92,13 +93,16 @@ def hg_log(repo_dir):
...
+        backed_out = False
+        if rev[5] != b'':
+            backed_out = True
         revs.append(Commit(
             node=rev[0],
             author=rev[1],
             desc=rev[2],
             date=dt,
             bug=rev[4],
+            ever_backedout=backed_out,
```
Comment (marco-c) [inline bugbug/repository.py]: You can just do `ever_backedout=(rev[5] != b'')` and avoid defining the backed_out variable.

File: bugbug/repository.py [hunk 4]
```diff
@@ -92,13 +93,13 @@ def hg_log(repo_dir):
     for rev in hglib.util.grouper(template.count('\\0'), out):
         posixtime = float(rev[3].split(b'.', 1)[0])
         dt = datetime.fromtimestamp(posixtime)
-
```
Comment (marco-c) [inline bugbug/repository.py]: Don't remove this newline (better not to change things that are unrelated to the goal of the patch).

File: bugbug/bug_features.py [hunk 5]
```diff
@@ -230,6 +230,11 @@ def __call__(self, bug, **kwargs):
         return sum(res) / len(res)
 
 
+class commit_no_of_backouts(object):
+    def __call__(self, bug, **kwargs):
+        return sum(1 for commit in bug['commits'] if commit['ever_backedout'] is True)
```
Comment (marco-c) [inline bugbug/bug_features.py]: Forgot to tell you before, but here you don't need `is True`

Comment (marco-c) [reply bugbug/bug_features.py]: `return sum(1 for commit in bug['commits'] if commit['ever_backedout'])`

Comment (ayir) [reply bugbug/bug_features.py]: Ohh thanks!Just removed it.

File: bugbug/repository.py [hunk 6]
```diff
@@ -91,14 +92,15 @@ def hg_log(repo_dir):
     revs = []
     for rev in hglib.util.grouper(template.count('\\0'), out):
         posixtime = float(rev[3].split(b'.', 1)[0])
+
```
Comment (marco-c) [inline bugbug/repository.py]: You added a newline here instead of readding the one you removed :smile:

Comment (ayir) [reply bugbug/repository.py]: Just a minute. sorry
